### PR TITLE
CALCITE-2815 GeodeSchemaFactory change to pass in clientCache using geode jndi context

### PR DESCRIPTION
**Pull request by chadasa**

### CODE REVIEW

**Comment by asereda-gs** [comment]
Pls rebase all commits into one

File: geode/src/main/java/org/apache/calcite/adapter/geode/rel/GeodeSchemaFactory.java [hunk 1]
```diff
@@ -39,18 +45,58 @@
   public static final String PDX_SERIALIZABLE_PACKAGE_PATH = "pdxSerializablePackagePath";
   public static final String ALLOW_SPATIAL_FUNCTIONS = "spatialFunction";
   public static final String COMMA_DELIMITER = ",";
+  public static final String GEODE_JNDI_INITIAL_CONTEXT_FACTORY =
+      "org.apache.geode.internal.jndi.InitialContextFactoryImpl";
```
Comment (asereda-gs) [inline geode/src/main/java/org/apache/calcite/adapter/geode/rel/GeodeSchemaFactory.java]: Why do we depend on geode internal API ? `geode.internal.jndi.InitialContextFactoryImpl`

File: geode/src/main/java/org/apache/calcite/adapter/geode/rel/GeodeSchemaFactory.java [hunk 2]
```diff
@@ -39,18 +45,58 @@
...
   }
 
+  private static Context getContext() {
+    Context context = null;
+
+    try {
+      Hashtable env = new Hashtable();
+      env.put(Context.INITIAL_CONTEXT_FACTORY, GEODE_JNDI_INITIAL_CONTEXT_FACTORY);
+      context = new InitialContext(env);
+    } catch (Exception ignored) {
```
Comment (asereda-gs) [inline geode/src/main/java/org/apache/calcite/adapter/geode/rel/GeodeSchemaFactory.java]: can we just catch [NamingException](https://docs.oracle.com/javase/7/docs/api/javax/naming/NamingException.html) here ?

File: geode/src/main/java/org/apache/calcite/adapter/geode/rel/GeodeSchemaFactory.java [hunk 3]
```diff
@@ -39,18 +45,58 @@
...
+        String[] regionNames = ((String) map.get(REGIONS)).split(COMMA_DELIMITER);
+        String pbxSerializablePackagePath = (String) map.get(PDX_SERIALIZABLE_PACKAGE_PATH);
+
+        gemFireCache = GeodeUtils.createClientCache(locatorHost, locatorPort,
+            pbxSerializablePackagePath, true);
+      } else {
+        Context context = getContext();
+        gemFireCache = (GemFireCache) context.lookup(geodeClientCacheObjectKey);
+      }
+    } catch (Exception ignored) {
```
Comment (asereda-gs) [inline geode/src/main/java/org/apache/calcite/adapter/geode/rel/GeodeSchemaFactory.java]: I don't like ignoring coarse exceptions (as in `java.lang.Exception`). Can it be just [NamingException](https://docs.oracle.com/javase/7/docs/api/javax/naming/NamingException.html) ?

File: geode/src/test/java/org/apache/calcite/adapter/geode/rel/RelationalJdbcExample.java [hunk 4]
```diff
@@ -94,10 +81,81 @@ public static void main(String[] args) throws Exception {
...
+      break;
+    case 1:
+      runExampleWithJndiClientCacheObject();
+      break;
+    default:
+      throw new AssertionError("unknown example " + i);
+    }
+  }
+
+  public static void main(String[] args) throws Exception {
```
Comment (asereda-gs) [inline geode/src/test/java/org/apache/calcite/adapter/geode/rel/RelationalJdbcExample.java]: There is no way to write unit tests for that ? 
I don't think this class should have `main` method.

File: geode/src/test/java/org/apache/calcite/adapter/geode/rel/RelationalJdbcExample.java [hunk 5]
```diff
@@ -94,10 +81,81 @@ public static void main(String[] args) throws Exception {
...
+        "inline:"
+            + "{\n"
+            + "  version: '1.0',\n"
+            + "  schemas: [\n"
+            + "     {\n"
+            + "       type: 'custom',\n"
+            + "       name: 'TEST',\n"
+            + "       factory: 'org.apache.calcite.adapter.geode.rel.GeodeSchemaFactory',\n"
+            + "       operand: {\n"
+            + "         geodeClientCacheObjectKey: 'java:/TestClientCacheObject', \n"
```
Comment (asereda-gs) [inline geode/src/test/java/org/apache/calcite/adapter/geode/rel/RelationalJdbcExample.java]: If we agree that SchemaFactories should allow retrieval of objects using JNDI (not yet decided) then they should have consistent name for Elastic, Geode, Mongo :  `jndi-name` or `jndiName`  etc.

File: geode/src/main/java/org/apache/calcite/adapter/geode/rel/GeodeSchemaFactory.java [hunk 6]
```diff
@@ -39,18 +45,58 @@
...
   public GeodeSchemaFactory() {
     // Do Nothing
   }
 
+  private static Context getContext() {
+    Context context = null;
+
+    try {
+      Hashtable env = new Hashtable();
+      env.put(Context.INITIAL_CONTEXT_FACTORY, GEODE_JNDI_INITIAL_CONTEXT_FACTORY);
```
Comment (asereda-gs) [inline geode/src/main/java/org/apache/calcite/adapter/geode/rel/GeodeSchemaFactory.java]: Hmm what is JNDI context is configured externally ?

File: geode/src/main/java/org/apache/calcite/adapter/geode/rel/GeodeSchemaFactory.java (hunk 1 above)
Comment (chadasa) [reply geode/src/main/java/org/apache/calcite/adapter/geode/rel/GeodeSchemaFactory.java]: We would want to use ModelHandler call or just set database connection property **model**. This way the calcite connection details can still be passed as file or json string (with **inline:** prefixed) .
Will remove hard dependency of GEODE JNDI INTERNAL CONTEXT class

File: geode/src/test/java/org/apache/calcite/adapter/geode/rel/RelationalJdbcExample.java (hunk 4 above)
Comment (chadasa) [reply geode/src/test/java/org/apache/calcite/adapter/geode/rel/RelationalJdbcExample.java]: This class already has a main method, its an example class(All example classes have main methods).
Do you want me create a new class RelationalJdbcExampleRunner.class and have all the code move into it.
The main will call this RelationalJdbcExampleRunner class.
We can then add tests to RelationalJdbcExampleRunner.class

File: geode/src/test/java/org/apache/calcite/adapter/geode/rel/RelationalJdbcExample.java (hunk 5 above)
Comment (chadasa) [reply geode/src/test/java/org/apache/calcite/adapter/geode/rel/RelationalJdbcExample.java]: Will stick to camel case naming convention **testClientCacheObjectKey**

File: geode/src/main/java/org/apache/calcite/adapter/geode/rel/GeodeSchemaFactory.java (hunk 6 above)
Comment (chadasa) [reply geode/src/main/java/org/apache/calcite/adapter/geode/rel/GeodeSchemaFactory.java]: Will make change to pass HashTable env object of Context as part of json calcite model property

**Pull request closed without merging**